### ISSUE TITLE
converted-bert  fp16 model is slow  

### ISSUE THREAD

**Issue by zirui** [microsoft/onnxruntime#6060]
**Describe the bug**

I followed the [onnx-export  tutorial](https://github.com/huggingface/transformers/blob/master/notebooks/04-onnx-export.ipynb) to convert my `sentence pair` classification(2-class) model to onnx format, but i'm facing problems:
1. the onnx model **output different** logits
2. **float16** model is more than twice as **slow** than default(**float 32**) model

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 14.04
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version:1.4.0
- Python version: 3.6.5
- CUDA/cuDNN version: cuda-10.1 cudnn-8.04
- GPU model and memory: NVIDIA Tesla P40 24GB

**To Reproduce**

```bash
# conver to onnx model
python -m transformers.convert_graph_to_onnx --framework pt --model $input_model $output_model_path --pipeline sentiment-analysis

# optmized for bert model
python -m onnxruntime_tools.optimizer_cli \
       --input $output_model_path \
       --output $opt_output_model_path \
       --model_type bert \
       --hidden_size 1024 \
       --num_heads 16 \
       --use_gpu \
       --float16
```

below is the code-snippet of my test code:
```python

def create_model_for_provider(model_path: str, provider: str) -> InferenceSession:
    options = SessionOptions()
    options.graph_optimization_level = GraphOptimizationLevel.ORT_ENABLE_ALL
    session = InferenceSession(model_path, options, providers=[provider])
    session.disable_fallback()
    return session

def run_onnx_model(model, validation_dataloader, torch_model, args):
    count = 1
    pred_logits, ex_ids = [], []
    for idx, batch in tqdm.tqdm(
        enumerate(validation_dataloader),
        desc="eval batch",
        total=len(validation_dataloader),
    ):
        batch = tuple(t.to(args.device) for t in batch)
        #keys = ["input_ids", "attention_mask","token_type_ids", "labels"]
        keys = ["input_ids", "attention_mask","token_type_ids"]

        # inputs for pytorch/onnx
        model_inputs = dict(zip(keys, batch[:3]))

        inputs_onnx = {k: v.cpu().detach().numpy() for k, v in model_inputs.items()}

        # inference with torch model
        output_pytorch = torch_model(**model_inputs)

        print("pytorch output")
        print(output_pytorch)

        # inference with onnx model
        pooled = model.run(None, inputs_onnx)[0]
        logits = pooled[0]

        print("onnx outoput")
        print(pooled)

        pred_logits.append(logits)
        ex_ids.append(batch[-1])
    return pred_logits, ex_ids

def test(args):
    args = get_args()
    torch_model = BertForSequenceClassification.from_pretrained(args.pretrained_model_path, config=config)
    onnx_model = create_model_for_provider("onnx/roberta.onnx", "CUDAExecutionProvider")
    # create `sentence pair` features for classification
    validation_dataloader = get_input_features(text, args)
    logits, ex_ids = run_onnx_model(onnx_model, validation_dataloader, torch_model, args)
```

**output**
pytorch output
```
[array([[ 2.3372207, -2.3931568],
       [ 2.3980954, -2.2760265],
       [ 2.4161563, -2.3171213],
       [ 2.4139354, -2.3049214],
       [ 2.3819523, -2.2124367],
       [ 2.3976293, -2.3438153],
       [ 2.3811698, -2.3515754],
       [ 2.1626067, -2.1989098],
       [ 2.3606634, -2.399568 ],
       [ 2.361903 , -2.4018219],
       [ 2.3726542, -2.386125 ],
       [ 1.4030353, -1.2204564],
       [ 2.4067426, -2.4028478],
       [ 2.3399456, -2.349934 ],
       [ 2.4034119, -2.392361 ],
       [ 2.2462108, -2.1991572],
       [ 2.3456295, -2.3796468],
       [ 2.3579483, -2.390833 ],
       [ 2.4269786, -2.3824193],
       [ 2.2360866, -2.344134 ],
       [ 2.2120304, -2.2640295],
       [ 1.5615726, -1.2707071],
       [ 2.2761734, -2.3210094],
       [ 2.3432932, -2.3778472],
       [ 2.375787 , -2.3454647]], dtype=float32)]
```

onnx output
```
[[array([[ 2.4004993 , -2.128048  ],
       [ 1.8116348 , -1.0416967 ],
       [ 2.7683523 , -2.5471332 ],
       [ 0.48842308,  0.283095  ],
       [ 2.1037195 , -1.5077535 ],
       [ 2.144089  , -1.5617115 ],
       [ 3.075785  , -3.2294552 ],
       [ 2.4989674 , -2.2408962 ],
       [ 2.961293  , -2.9502418 ],
       [ 2.2245562 , -1.8007447 ],
       [ 3.1293359 , -3.5581563 ],
       [ 1.9777924 , -1.4636573 ],
       [ 3.05422   , -3.0377975 ],
       [ 2.9514942 , -3.1880434 ],
       [ 3.0133646 , -3.298267  ],
       [ 3.3254344 , -3.570813  ],
       [ 3.105355  , -3.1634367 ],
       [ 3.1248448 , -3.4561422 ],
       [ 3.2075515 , -3.546914  ],
       [ 3.0161104 , -3.296281  ],
       [ 3.0115733 , -3.3394403 ],
       [ 1.362302  , -0.70032716],
       [ 2.848809  , -3.0577686 ],
       [ 3.0407982 , -3.3116124 ],
       [ 2.9961717 , -3.3528812 ]], dtype=float32)]]
```

**performance comparsion**

|  model |inference-time   |
|---|---|
|  pytorch-gpu |  0.3483 |
|   onnxruntime-gpu|0.2881    |
|onnxruntime-gpu float16    | 0.6472  |

**Expected behavior**
1.  onnx model should output `same`(with small difference) as the pytorch model does
2. float16 model should be  faster than float32 model

**Comment by zirui** [comment]
the `different output` issue was caused by my silly mistake(missed used pytorch models), and i have solved it.
so i changed the title of this issue

**Comment by tianleiwu** [comment]
@zirui, I think the cause is M40 does not support FP16. 

You might try fp16 on P100, V100, T4 or newer GPUs. 

![image](https://user-images.githubusercontent.com/30328909/103990640-6c367180-5146-11eb-8ed1-95218efc8149.png)
If you look at spec of M40, it does not mention fp16. I think calculation will be done in FP32, and when model is FP16, there will be extra data conversion between FP32 and FP16. So, it is expected that FP16 is slower than FP32 in M40.

**Comment by zirui** [comment]
> @zirui, I think the cause is M40 does not support FP16.
> 
> You might try fp16 on P100, V100, T4 or newer GPUs.
> 
> ![image](https://user-images.githubusercontent.com/30328909/103990640-6c367180-5146-11eb-8ed1-95218efc8149.png)
> If you look at spec of M40, it does not mention fp16. I think calculation will be done in FP32, and when model is FP16, there will be extra data conversion between FP32 and FP16. So, it is expected that FP16 is slower than FP32 in M40.

thanks @tianleiwu,
I checked the gpu type of my test machine, actually it is Tesla P40(i have another machine with M40, and i confused them), and it seems that  neither M40 nor P40 support fp16,
i will close this issue
